### ISSUE TITLE
I broke memoization: Duplicate resource URN

### ISSUE THREAD

**Issue by pawelprazak** [VirtusLab/besom#165]
This:
```
val feedBucket = Output("Asdf").flatMap(_ =>
    s3.Bucket(
      "pulumi-catpost-cat-pics",
      s3.BucketArgs(
        forceDestroy = true
      )
    )
  )
...
for
    bucket <- feedBucket
    bucket <- feedBucket
  yield Pulumi.exports(
    feedBucket = bucket.bucket,
  )
```

Results in this error:
```
$ pulumi up -y
Enter your passphrase to unlock config/secrets
    (set PULUMI_CONFIG_PASSPHRASE or PULUMI_CONFIG_PASSPHRASE_FILE to remember):  
Enter your passphrase to unlock config/secrets
Previewing update (dev):
     Type                 Name                     Plan       Info
 +   pulumi:pulumi:Stack  besom-tutorial-dev       create     
 +   └─ aws:s3:Bucket     pulumi-catpost-cat-pics  create     1 error


Diagnostics:
  aws:s3:Bucket (pulumi-catpost-cat-pics):
    error: Duplicate resource URN 'urn:pulumi:dev::besom-tutorial::aws:s3/bucket:Bucket::pulumi-catpost-cat-pics'; try giving it a unique name
```

Known workaround id to make sure to never call resource creation twice but this might be tricky for the user to workaround this way in more complex programs.

**Comment by lbialy** [comment]
Fixed by #167

**Comment by pawelprazak** [comment]
Note: the discussion continues in #168
